File: zj_album/src/main/java/com/zj/album/ui/views/BaseLoadingView.java
package com.zj.album.ui.views;

import android.animation.Animator;
import android.animation.ArgbEvaluator;
import android.animation.ValueAnimator;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;
import android.support.v4.content.ContextCompat;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.ProgressBar;
import android.widget.TextView;
import com.zj.album.R;
import org.jetbrains.annotations.Nullable;

import java.util.HashMap;
import java.util.Map;

/**
 * Created by ZJJ on 2018/7/3.
 */

@SuppressWarnings({"unused", "unchecked"})
public class BaseLoadingView extends FrameLayout {

    public BaseLoadingView(Context context) {
        this(context, null, 0);
    }

    public BaseLoadingView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public BaseLoadingView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context, attrs);
    }

    private static final int defaultAnimationDuration = 400;

    private DisplayMode oldMode = DisplayMode.none;
    private Map<DisplayMode, Float> disPlayViews;
    private View rootView;
    private View noData, noNetwork;
    private ProgressBar loading;
    private TextView tvHint, tvRefresh;
    private CallRefresh refresh;
    private int bgColor;
    private int bgColorOnAct;
    private int needBackgroundColor, oldBackgroundColor;
    private int noDataRes = -1;
    private int noNetworkRes = -1;
    private int loadingRes = -1;
    private int hintTextColor, refreshTextColor;
    private ArgbEvaluator argbEvaluator;
    private boolean refreshEnable = true;
    private boolean refreshEnableWithView = false;

    private BaseLoadingValueAnimator valueAnimator;

    public void setRefreshEnable(boolean enable) {
        this.refreshEnable = enable;
    }

    public interface CallRefresh {
        void onCallRefresh();
    }

    public enum DisplayMode {
        none(0), loading(1), noData(2), noNetwork(3), normal(4);

        private final int value;

        DisplayMode(int value) {
            this.value = value;
        }
    }

    /**
     * when you set mode as noData/noNetwork ,
     * you can get the event when this view was clicked
     * and you can refresh content  when the  "onCallRefresh()" callback
     */
    public void setRefreshListener(CallRefresh refresh) {
        this.refresh = refresh;
        setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                if (refreshEnable && refreshEnableWithView && BaseLoadingView.this.refresh != null) {
                    BaseLoadingView.this.refresh.onCallRefresh();
                }
            }
        });
    }

    private void init(Context context, AttributeSet attrs) {
        if (attrs != null) {
            @Nullable TypedArray array = context.obtainStyledAttributes(attrs, R.styleable.BaseLoadingView);
            try {
                bgColor = array.getColor(R.styleable.BaseLoadingView_backgroundFill, ContextCompat.getColor(context, R.color.loading_color_background));
                bgColorOnAct = array.getColor(R.styleable.BaseLoadingView_backgroundOnAct, ContextCompat.getColor(context, R.color.loading_color_background_float));
                noDataRes = array.getResourceId(R.styleable.BaseLoadingView_noDataRes, -1);
                noNetworkRes = array.getResourceId(R.styleable.BaseLoadingView_noNetworkRes, -1);
                loadingRes = array.getResourceId(R.styleable.BaseLoadingView_loadingRes, -1);
                hintTextColor = array.getColor(R.styleable.BaseLoadingView_hintColor, -1);
                refreshTextColor = array.getColor(R.styleable.BaseLoadingView_refreshTextColor, -1);
            } catch (Exception ignore) {
            } finally {
                array.recycle();
            }
        }
        initView(context);
    }

    private void initView(Context context) {
        rootView = inflate(context, R.layout.loading_view, this);
        noData = f(R.id.blv_vNoData);
        noNetwork = f(R.id.blv_vNoNetwork);
        loading = f(R.id.blv_pb);
        tvHint = f(R.id.blv_tvHint);
        tvRefresh = f(R.id.blv_tvRefresh);
        disPlayViews = new HashMap<>();
        if (hintTextColor > 0)
            tvHint.setTextColor(hintTextColor);
        if (refreshTextColor > 0)
            tvRefresh.setTextColor(refreshTextColor);
        resetUi();
        argbEvaluator = new ArgbEvaluator();
    }

    private BaseLoadingAnimatorListener listener = new BaseLoadingAnimatorListener() {

        @Override
        public void onDurationChange(ValueAnimator animation, float offset, DisplayMode mode, boolean isShowOnAct) {
            synchronized (BaseLoadingView.this) {
                onAnimationFraction(animation.getAnimatedFraction(), offset, mode);
            }
        }

        @Override
        public void onAnimEnd(Animator animation, DisplayMode mode, boolean isShowOnAct) {
            synchronized (BaseLoadingView.this) {
                onAnimationFraction(1.0f, 1.0f, mode);
            }
        }
    };

    /**
     * @param drawableRes must be an animatorDrawable in progressBar;
     * @link call resetUi() after set this
     */
    public BaseLoadingView setLoadingDrawable(int drawableRes) {
        this.loadingRes = drawableRes;
        return this;
    }

    //call resetUi() after set this
    public BaseLoadingView setNoDataDrawable(int drawableRes) {
        this.noDataRes = drawableRes;
        return this;
    }

    //call resetUi() after set this
    public BaseLoadingView setNoNetworkDrawable(int drawableRes) {
        this.noNetworkRes = drawableRes;
        return this;
    }

    //reset loading/noData/noNetwork drawable
    private void resetUi() {
        if (loadingRes > 0) {
            Drawable drawable = ContextCompat.getDrawable(getContext(), loadingRes);
            if (drawable != null) {
                Rect rect = loading.getIndeterminateDrawable().getBounds();
                drawable.setBounds(rect);
                loading.setIndeterminateDrawable(drawable);
            }
        }
        if (noDataRes > 0) {
            noData.setBackgroundResource(noDataRes);
        }
        if (noNetworkRes > 0) {
            noNetwork.setBackgroundResource(noNetworkRes);
        }
    }

    /**
     * just call setMode after this View got,
     *
     * @param mode      the current display mode you need;
     * @param showOnAct is showing on content? or hide content?
     * @param hint      show something when it`s change a mode;
     */
    public void setMode(DisplayMode mode, String hint, boolean showOnAct) {
        if (mode == DisplayMode.none) mode = DisplayMode.normal;
        int hashCode = (showOnAct ? -10 : 10) + mode.value;
        int curCode = (showOnAct ? -10 : 10) + oldMode.value;
        oldMode = mode;
        boolean isSameMode = hashCode == curCode;

        String hintText = (!TextUtils.isEmpty(hint) ? hint : getHintString(mode));
        if (hint != null) {
            tvHint.setText(hint);
        }
        refreshEnableWithView = refreshEnable && (mode == DisplayMode.noData || mode == DisplayMode.noNetwork);
        tvRefresh.setVisibility(refreshEnableWithView ? View.VISIBLE : View.GONE);
        if (valueAnimator == null) {
            valueAnimator = new BaseLoadingValueAnimator(listener);
            valueAnimator.setDuration(defaultAnimationDuration);
        } else {
            valueAnimator.end();
        }
        disPlayViews.put(mode, 0.0f);
        //背景样式改变或mode改变，需要重绘背景或样式
        if (!isSameMode) {
            needBackgroundColor = showOnAct ? bgColorOnAct : bgColor;
            valueAnimator.start(mode, showOnAct);
        }
    }

    private String getHintString(DisplayMode mode) {
        switch (mode) {
            case loading:
                return getContext().getString(R.string.loading_progress);
            case noData:
                return getContext().getString(R.string.loading_no_data);
            case noNetwork:
                return getContext().getString(R.string.loading_no_network);
            default:
                return "";
        }
    }

    /**
     * just call setMode after this View got,
     *
     * @param mode      the current display mode you need;
     * @param showOnAct is showing on content? or hide content?
     * @param hint      show something when it`s change a mode;
     */
    public void setMode(DisplayMode mode, String hint, boolean showOnAct, int delayDismissTime) {
        setMode(mode, hint, showOnAct);
        postDelayed(new Runnable() {
            @Override
            public void run() {
                setMode(DisplayMode.normal, "", false);
            }
        }, delayDismissTime);
    }


    private synchronized void onAnimationFraction(float duration, float offset, DisplayMode curMode) {
        setViews(offset, curMode);
        setBackground(duration, offset, curMode);
    }


    private void setViews(float offset, DisplayMode curMode) {
        for (Map.Entry<DisplayMode, Float> entry : disPlayViews.entrySet()) {
            View curSetView = getDisplayView(entry.getKey());
            if (curSetView != null) {
                float curAlpha = entry.getValue();
                float newAlpha;
                if (entry.getKey() == curMode) {
                    //need show
                    if (curSetView.getVisibility() != VISIBLE) {
                        curSetView.setVisibility(VISIBLE);
                        curSetView.setAlpha(0);
                    }
                    newAlpha = Math.min(1.0f, Math.max(0.0f, curAlpha) + offset);
                    curSetView.setAlpha(newAlpha);
                } else {
                    //need hide
                    newAlpha = Math.max(Math.min(1.0f, curAlpha) - offset, 0);
                    curSetView.setAlpha(newAlpha);
                    if (newAlpha == 0 && curSetView.getVisibility() != GONE)
                        curSetView.setVisibility(GONE);
                }
                disPlayViews.put(entry.getKey(), newAlpha);
            }
        }
    }

    private void setBackground(float duration, float offset, DisplayMode curMode) {
        if (curMode != DisplayMode.normal) {
            //画背景
            if (getVisibility() != VISIBLE) {
                setVisibility(VISIBLE);
                setAlpha(0);
            }
            if (getAlpha() >= 1.0f) {
                if (oldBackgroundColor != needBackgroundColor) {
                    int curBackgroundColor = (int) argbEvaluator.evaluate(duration, oldBackgroundColor, needBackgroundColor);
                    oldBackgroundColor = curBackgroundColor;
                    setBackgroundColor(curBackgroundColor);
                }
            } else {
                setAlpha(Math.min(1.0f, duration));
                if (oldBackgroundColor != needBackgroundColor) {
                    setBackgroundColor(needBackgroundColor);
                    oldBackgroundColor = needBackgroundColor;
                }
            }
        } else {
            setAlpha(1.0f - duration);
            if (getAlpha() <= 0.05f) {
                setAlpha(0);
                setBackgroundColor(oldBackgroundColor = 0);
                setVisibility(GONE);
            }
        }
    }

    private View getDisplayView(DisplayMode mode) {
        switch (mode) {
            case noData:
                return noData;
            case loading:
                return loading;
            case noNetwork:
                return noNetwork;
        }
        return null;
    }

    private <T extends View> T f(int id) {
        return (T) rootView.findViewById(id);
    }


    public static class BaseLoadingValueAnimator extends ValueAnimator {

        private DisplayMode curMode;
        private boolean isShowOnAct;
        private float curDuration;
        private boolean isCancel;

        private BaseLoadingAnimatorListener listener;

        private void start(DisplayMode mode, boolean isShowOnAct) {
            if (isRunning()) cancel();
            this.curMode = mode;
            this.isShowOnAct = isShowOnAct;
            super.start();
        }

        @Override
        public void cancel() {
            removeAllListeners();
            if (listener != null) listener = null;
            isCancel = true;
            super.cancel();
        }

        private BaseLoadingValueAnimator(BaseLoadingAnimatorListener l) {
            this.listener = l;
            setFloatValues(0.0f, 1.0f);
            addListener(new AnimatorListener() {
                @Override
                public void onAnimationStart(Animator animation) {
                    if (curDuration != 0) curDuration = 0;
                }

                @Override
                public void onAnimationEnd(Animator animation) {
                    curDuration = 0;
                    if (isCancel) return;
                    if (listener != null)
                        listener.onAnimEnd(animation, curMode, isShowOnAct);
                }

                @Override
                public void onAnimationCancel(Animator animation) {
                    curDuration = 0;
                }

                @Override
                public void onAnimationRepeat(Animator animation) {
                    curDuration = 0;
                }
            });

            addUpdateListener(new AnimatorUpdateListener() {
                @Override
                public void onAnimationUpdate(ValueAnimator animation) {
                    if (isCancel) return;
                    if (listener != null) {
                        float duration = (float) animation.getAnimatedValue();
                        float offset = duration - curDuration;
                        listener.onDurationChange(animation, offset, curMode, isShowOnAct);
                        curDuration = duration;
                    }
                }
            });
        }

        public void setAnimatorListener(BaseLoadingAnimatorListener listener) {
            this.listener = listener;
        }
    }

    public interface BaseLoadingAnimatorListener {

        void onDurationChange(ValueAnimator animation, float duration, DisplayMode mode, boolean isShowOnAct);

        void onAnimEnd(Animator animation, DisplayMode mode, boolean isShowOnAct);
    }
}

File: zj_album/src/main/java/com/zj/album/ui/views/BaseLoadingView.kt
@file:Suppress("unused")

package com.zj.album.ui.views

import android.animation.Animator
import android.animation.ArgbEvaluator
import android.animation.ValueAnimator
import android.content.Context
import androidx.core.content.ContextCompat
import android.text.TextUtils
import android.util.AttributeSet
import android.view.View
import android.widget.FrameLayout
import android.widget.ProgressBar
import android.widget.TextView
import com.zj.album.R
import java.util.*
import kotlin.math.max
import kotlin.math.min

/**
 * @author ZJJ on 2018/7/3.
 */
internal class BaseLoadingView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) : FrameLayout(context, attrs, defStyleAttr) {

    private var oldMode = DisplayMode.NONE
    private var disPlayViews: MutableMap<DisplayMode, Float>? = null
    private var contentView: View? = null
    private var vLoading: ProgressBar? = null
    private var vNoData: View? = null
    private var vNoNetWork: View? = null
    private var blvChildBg: View? = null
    private var curBackgroundView: View? = null
    private var tvHint: TextView? = null
    private var tvRefresh: TextView? = null
    private var refresh: (() -> Unit)? = null

    private var bgColor: Int = 0
    private var bgColorOnAct: Int = 0
    private var needBackgroundColor: Int = 0
    private var oldBackgroundColor: Int = 0
    private var noDataRes = -1
    private var noNetworkRes = -1
    private var loadingRes = -1
    private var hintTextColor: Int = 0
    private var refreshTextColor: Int = 0

    private var showOnActDefault: Boolean = false

    private var loadingHint: String = ""
    private var noDataHint: String = ""
    private var networkErrorHint: String = ""
    private var refreshHint: String? = ""

    private var argbEvaluator: ArgbEvaluator? = null
    private var refreshEnable = true
    private var refreshEnableWithView = false

    private var valueAnimator: BaseLoadingValueAnimator? = null

    private val listener = object : BaseLoadingAnimatorListener {

        override fun onDurationChange(animation: ValueAnimator, duration: Float, mode: DisplayMode?, isShowOnAct: Boolean) {
            synchronized(this@BaseLoadingView) {
                onAnimationFraction(animation.animatedFraction, duration, mode)
            }
        }

        override fun onAnimEnd(animation: Animator, mode: DisplayMode?, isShowOnAct: Boolean) {
            synchronized(this@BaseLoadingView) {
                onAnimationFraction(1.0f, 1.0f, mode)
            }
        }
    }

    init {
        init(context, attrs)
    }

    fun setRefreshEnable(enable: Boolean) {
        this.refreshEnable = enable
    }

    enum class DisplayMode(internal val value: Int) {
        NONE(0), LOADING(1), NO_DATA(2), NET_ERROR(3), DISMISS(4)
    }

    /**
     * when you set mode as NO_DATA/NET_ERROR ,
     * you can get the event when this view was clicked
     * and you can refresh content  when the  "onCallRefresh()" callback
     */
    fun setRefreshListener(refresh: () -> Unit) {
        this.refresh = refresh
        setOnClickListener {
            if (refreshEnable && refreshEnableWithView && this@BaseLoadingView.refresh != null) {
                this@BaseLoadingView.refresh?.invoke()
            }
        }
    }

    private fun init(context: Context, attrs: AttributeSet?) {
        if (attrs != null) {
            val array = context.obtainStyledAttributes(attrs, R.styleable.BaseLoadingView)
            try {
                bgColor = array.getColor(R.styleable.BaseLoadingView_backgroundFill, ContextCompat.getColor(context, R.color.pg_color_loading_color_background))
                bgColorOnAct = array.getColor(R.styleable.BaseLoadingView_backgroundOnAct, ContextCompat.getColor(context, R.color.pg_color_loading_background_float))
                noDataRes = array.getResourceId(R.styleable.BaseLoadingView_noDataRes, -1)
                noNetworkRes = array.getResourceId(R.styleable.BaseLoadingView_noNetworkRes, -1)
                loadingRes = array.getResourceId(R.styleable.BaseLoadingView_loadingRes, -1)
                hintTextColor = array.getColor(R.styleable.BaseLoadingView_hintColor, -1)
                refreshTextColor = array.getColor(R.styleable.BaseLoadingView_refreshTextColor, -1)
                loadingHint = array.getString(R.styleable.BaseLoadingView_loadingText) ?: ""
                noDataHint = array.getString(R.styleable.BaseLoadingView_noDataText) ?: ""
                networkErrorHint = array.getString(R.styleable.BaseLoadingView_networkErrorText) ?: ""
                refreshHint = array.getString(R.styleable.BaseLoadingView_refreshText)
                showOnActDefault = array.getBoolean(R.styleable.BaseLoadingView_showOnActDefault, false)
                refreshEnable = array.getBoolean(R.styleable.BaseLoadingView_refreshEnable, true)
            } catch (e: Exception) {
                e.printStackTrace()
            } finally {
                array.recycle()
            }
        }
        initView(context)
    }

    private fun initView(context: Context) {
        contentView = View.inflate(context, R.layout.loading_view, this)
        vNoData = f(R.id.blv_vNoData)
        vNoNetWork = f(R.id.blv_vNoNetwork)
        vLoading = f(R.id.blv_pb)
        tvHint = f(R.id.blv_tvHint)
        tvRefresh = f(R.id.blv_tvRefresh)
        blvChildBg = f(R.id.blv_child_bg)
        if (refreshHint != null && !refreshHint.isNullOrEmpty()) tvRefresh?.text = refreshHint
        if (hintTextColor != 0) tvHint?.setTextColor(hintTextColor)
        if (refreshTextColor != 0) tvRefresh?.setTextColor(refreshTextColor)
        argbEvaluator = ArgbEvaluator()
        disPlayViews = EnumMap(com.zj.album.ui.views.BaseLoadingView.DisplayMode::class.java)
        disPlayViews?.put(DisplayMode.LOADING, 0.0f)
        tvHint?.text = loadingHint
        resetUi()
        resetBackground(showOnActDefault)
    }

    private fun resetBackground(showOnAct: Boolean) {
        curBackgroundView = if (showOnAct) blvChildBg else this
        blvChildBg?.setBackgroundColor(if (showOnAct) bgColorOnAct else 0)
        setBackgroundColor(if (showOnAct) 0 else bgColor)
    }

    /**
     * @param drawableRes must be an animatorDrawable in progressBar;
     * @link call resetUi() after set this
     */
    fun setLoadingDrawable(drawableRes: Int): BaseLoadingView {
        this.loadingRes = drawableRes
        return this
    }

    //call resetUi() after set this
    fun setNoDataDrawable(drawableRes: Int): BaseLoadingView {
        this.noDataRes = drawableRes
        return this
    }

    //call resetUi() after set this
    fun setNetErrorDrawable(drawableRes: Int): BaseLoadingView {
        this.noNetworkRes = drawableRes
        return this
    }

    //reset LOADING/NO_DATA/NET_ERROR drawable
    private fun resetUi() {
        if (loadingRes > 0) {
            val drawable = context.getDrawable(loadingRes)
            if (drawable != null) {
                val rect = vLoading?.indeterminateDrawable?.bounds
                if (rect != null) drawable.bounds = rect
                vLoading?.indeterminateDrawable = drawable
            }
        }
        if (noDataRes > 0) {
            vNoData?.setBackgroundResource(noDataRes)
        }
        if (noNetworkRes > 0) {
            vNoNetWork?.setBackgroundResource(noNetworkRes)
        }
    }

    /**
     * just call setMode after this View got,
     *
     * @param m      the current display mode you need;
     * @param showOnContent is showing on content? or hide content?
     * @param hint      show something when it`s change a mode;
     */
    @JvmOverloads
    fun setMode(m: DisplayMode, hint: String = "", showOnContent: Boolean? = null) {
        var mode = m
        var showOnAct = showOnContent
        if (showOnAct == null) showOnAct = showOnActDefault
        if (mode == DisplayMode.NONE) mode = DisplayMode.DISMISS
        val newCode = (if (showOnAct) -10 else 10) + mode.value
        val oldCode = (if (showOnAct) -10 else 10) + oldMode.value
        oldMode = mode
        val isSameMode = newCode == oldCode
        val hintText = if (!TextUtils.isEmpty(hint)) hint else getHintString(mode)
        if (hintText.isNotEmpty()) {
            tvHint?.text = hintText
        }
        refreshEnableWithView = refreshEnable && (mode == DisplayMode.NO_DATA || mode == DisplayMode.NET_ERROR)
        tvRefresh?.visibility = if (refreshEnableWithView) View.VISIBLE else View.INVISIBLE
        if (valueAnimator == null) {
            valueAnimator = BaseLoadingValueAnimator(listener)
            valueAnimator?.duration = DEFAULT_ANIM_DURATION
        } else {
            valueAnimator?.end()
        }
        disPlayViews?.put(mode, 0.0f)
        if (!isSameMode) {
            resetBackground(showOnAct)
            needBackgroundColor = if (showOnAct) bgColorOnAct else bgColor
            valueAnimator?.start(mode, showOnAct)
        }
    }

    private fun getHintString(mode: DisplayMode): String {
        return when (mode) {
            DisplayMode.LOADING -> if (loadingHint.isEmpty()) "LOADING" else loadingHint
            DisplayMode.NO_DATA -> if (noDataHint.isEmpty()) "no data found" else noDataHint
            DisplayMode.NET_ERROR -> if (networkErrorHint.isEmpty()) "no network access" else networkErrorHint
            else -> ""
        }
    }

    fun hideDelay(delayDismissTime: Int) {
        postDelayed({ setMode(DisplayMode.DISMISS, "", false) }, delayDismissTime.toLong())
    }


    @Synchronized
    private fun onAnimationFraction(duration: Float, offset: Float, curMode: DisplayMode?) {
        setViews(offset, curMode)
        setBackground(duration, curMode)
    }


    private fun setViews(offset: Float, curMode: DisplayMode?) {
        disPlayViews?.forEach { (key, curAlpha) ->
            val curSetView = getDisplayView(key)
            if (curSetView != null) {
                val newAlpha: Float
                if (key == curMode) {
                    //need show
                    if (curSetView.visibility != View.VISIBLE) {
                        curSetView.visibility = View.VISIBLE
                        curSetView.alpha = 0f
                    }
                    newAlpha = min(1.0f, max(0.0f, curAlpha) + offset)
                    curSetView.alpha = newAlpha
                } else {
                    //need hide
                    newAlpha = max(min(1.0f, curAlpha) - offset, 0f)
                    curSetView.alpha = newAlpha
                    if (newAlpha == 0f && curSetView.visibility != View.GONE) curSetView.visibility = View.GONE
                }
                disPlayViews?.put(key, newAlpha)
            }
        }
    }

    private fun setBackground(duration: Float, curMode: DisplayMode?) {
        if (curMode != DisplayMode.DISMISS) {
            if (visibility != View.VISIBLE) {
                alpha = 0f
                visibility = View.VISIBLE
            }
            if (alpha >= 1.0f) {
                if (oldBackgroundColor != needBackgroundColor) {
                    curBackgroundView?.setBackgroundColor(needBackgroundColor)
                    oldBackgroundColor = needBackgroundColor
                }
            } else {
                alpha = min(1.0f, duration)
                if (oldBackgroundColor != needBackgroundColor) {
                    val curBackgroundColor = argbEvaluator?.evaluate(duration, oldBackgroundColor, needBackgroundColor) as Int
                    oldBackgroundColor = curBackgroundColor
                    curBackgroundView?.setBackgroundColor(curBackgroundColor)
                }
            }
        } else {
            alpha = 1.0f - duration
            if (alpha <= 0.05f) {
                alpha = 0f
                setBackgroundColor({ oldBackgroundColor = 0;oldBackgroundColor }.invoke())
                visibility = View.GONE
            }
        }
    }

    private fun getDisplayView(mode: DisplayMode): View? {
        return when (mode) {
            DisplayMode.NO_DATA -> vNoData
            DisplayMode.LOADING -> vLoading
            DisplayMode.NET_ERROR -> vNoNetWork
            else -> null
        }
    }

    private fun <T : View> f(id: Int): T? {
        return contentView?.findViewById<T>(id)
    }

    private class BaseLoadingValueAnimator constructor(private var listener: BaseLoadingAnimatorListener?) : ValueAnimator() {

        private var curMode: DisplayMode? = null
        private var isShowOnAct: Boolean = false
        private var curDuration: Float = 0.toFloat()
        private var isCancel: Boolean = false

        fun start(mode: DisplayMode, isShowOnAct: Boolean) {
            if (isRunning) cancel()
            this.curMode = mode
            this.isShowOnAct = isShowOnAct
            super.start()
        }

        override fun cancel() {
            removeAllListeners()
            if (listener != null) listener = null
            isCancel = true
            super.cancel()
        }

        init {
            setFloatValues(0.0f, 1.0f)
            addListener(object : AnimatorListener {
                override fun onAnimationStart(animation: Animator) {
                    if (curDuration != 0f) curDuration = 0f
                }

                override fun onAnimationEnd(animation: Animator) {
                    curDuration = 0f
                    if (isCancel) return
                    if (listener != null) listener?.onAnimEnd(animation, curMode, isShowOnAct)
                }

                override fun onAnimationCancel(animation: Animator) {
                    curDuration = 0f
                }

                override fun onAnimationRepeat(animation: Animator) {
                    curDuration = 0f
                }
            })

            addUpdateListener(AnimatorUpdateListener { animation ->
                if (isCancel) return@AnimatorUpdateListener
                if (listener != null) {
                    val duration = animation.animatedValue as Float
                    val offset = duration - curDuration
                    listener?.onDurationChange(animation, offset, curMode, isShowOnAct)
                    curDuration = duration
                }
            })
        }

        fun setAnimatorListener(listener: BaseLoadingAnimatorListener) {
            this.listener = listener
        }
    }

    interface BaseLoadingAnimatorListener {

        fun onDurationChange(animation: ValueAnimator, duration: Float, mode: DisplayMode?, isShowOnAct: Boolean)

        fun onAnimEnd(animation: Animator, mode: DisplayMode?, isShowOnAct: Boolean)
    }

    companion object {

        private const val DEFAULT_ANIM_DURATION = 400L
    }
}
